### PR TITLE
Implementation of validation error suppression via a Preprocessor

### PR DESCRIPTION
**Pull request by gerryhocks**
This is an implementation of a mechanism to support inline error suppression. This should work across all parser implementations. To suppress errors, place:

//@Suppress@ 

on a line. It should suppress validation errors for the section, block or paragraph it is placed before.

To suppress specific validations, add them to the line, separated by a space. For example:

//@Suppress@ spelling SuccessiveWord

These names are matched with the validators in a case-insensitive way.

### CODE REVIEW

**Comment by takahi-i** [comment]
👍

File: redpen-core/src/main/java/cc/redpen/RedPen.java [hunk 1]
```diff
@@ -138,7 +139,6 @@ public Document parse(DocumentParser parser, String content) throws RedPenExcept
     public Map<Document, List<ValidationError>> validate(List<Document> documents) {
         Map<Document, List<ValidationError>> docErrorsMap = new HashMap<>();
         documents.forEach(e -> docErrorsMap.put(e, new ArrayList<>()));
-        initializeValidators();
```
Comment (takahi-i) [inline redpen-core/src/main/java/cc/redpen/RedPen.java]: Thank you for the fix of UI @gerryhocks! But this line is needed to fix https://github.com/redpen-cc/redpen/issues/621

Comment (gerryhocks) [inline redpen-core/src/main/java/cc/redpen/RedPen.java]: Ah.. I see!

I think then I will need to get the UI to parse a fake document at startup to load the relevant validator configurations. Let me see how that goes.

Comment (takahi-i) [inline redpen-core/src/main/java/cc/redpen/RedPen.java]: Parsing a fake document in UI is a good idea.

Comment (gerryhocks) [inline redpen-core/src/main/java/cc/redpen/RedPen.java]: I've tried that and it seems to work okay - the pull request now reflects this change.

Comment (takahi-i) [inline redpen-core/src/main/java/cc/redpen/RedPen.java]: Thanks! The fix looks perfect.

**Comment by gerryhocks** [comment]
I've updated the pull request to restrict the suppress annotation to only be processed for AsciiDoc.

I found a small exception the the re:VIEW parser when it was accidentally given AsciiDoc to parse and guarded against it.

This request also includes the patch to make the redpen-server GUI work again, and includes some UI refreshes done in parallel with the other work.

I apologise for adding so many different things in this one single pull request.

File: redpen-core/src/main/java/cc/redpen/parser/PreprocessorRule.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    /**
+     * Return true if the rule is triggered by the given line and name given the structure
+     * of the supplied document
+     *
+     * @param document
+     * @param otherLineNumber
+     * @param name
+     * @return
+     */
+    public boolean isTriggeredBy(Document document, int otherLineNumber, String name) {
```
Comment (takahi-i) [inline redpen-core/src/main/java/cc/redpen/parser/PreprocessorRule.java]: I would like to have test cases including corner cases, to understand the expected behavior.

File: redpen-core/src/main/java/cc/redpen/parser/PreprocessingReader.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    public void close() throws IOException {
+        reader.close();
+    }
+
+    public String readLine() throws IOException {
+        String line = reader.readLine();
+        lineNumber++;
+        if (line != null) {
+            String ruleText = line;
+            if (parser instanceof AsciiDocParser) {
```
Comment (takahi-i) [inline redpen-core/src/main/java/cc/redpen/parser/PreprocessingReader.java]: 👍

File: redpen-core/src/main/java/cc/redpen/parser/PreprocessorRule.java (hunk 2 above)
Comment (gerryhocks) [inline redpen-core/src/main/java/cc/redpen/parser/PreprocessorRule.java]: Of course.. I'll add to and improve the test cases.

Comment (takahi-i) [inline redpen-core/src/main/java/cc/redpen/parser/PreprocessorRule.java]: Thanks!

Comment (gerryhocks) [inline redpen-core/src/main/java/cc/redpen/parser/PreprocessorRule.java]: I've pushed a change to the test case to try to highlight how isTriggeredBy works. Please let me know if you require more tests.

Comment (takahi-i) [inline redpen-core/src/main/java/cc/redpen/parser/PreprocessorRule.java]: Thanks!

**Comment by takahi-i** [comment]
LGTM
